[Star Tree] [Search] Keyword & Numeric Terms Aggregation

sandeshkr419 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
- Adds support for keyword & numeric terms aggregation.
- Aggregations are supported with or without metric sub-aggregations
- Aggregations are supported with previously supported queries for filtering (no change in support of adding up queries to requests)
- ~Fix for timestamp field to be fetched from request/valueSource instead of hard-coded value~ (this change is already merged https://github.com/opensearch-project/OpenSearch/pull/17239)

### Related Issues
Resolves #16551

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

bharath-techie commented on server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeQueryHelper.java:
```diff
@@ -146,6 +157,34 @@ private static boolean validateStarTreeMetricSupport(
+        AggregatorFactory aggregatorFactory
+    ) {
+        if (!(aggregatorFactory instanceof TermsAggregatorFactory termsAggregatorFactory)
+            || aggregatorFactory.getSubFactories().getFactories().length < 1) {
```
We can revisit this later post 2.19 - in my benchmarks , i see good improvements even without sub aggs for global terms aggs. For numeric terms, since there is no short-cut optimizations, we should always use star tree.

sandeshkr419 replied:
I'm keeping this PR contained with fully tested query shapes as of now. Will keep a TODO to evaluate bucket aggs with sub-aggs post this change.

bharath-techie replied:
yeah agreed. The ask is only for post 2.19 , we can have data via benchmarks and take the call for terms with no sub aggs. But for numeric terms, we definitely must do it since its guaranteed to be faster as the original collector has no optimizations. 

In my benchmarks it was anywhere between 2x and 5x faster or more for both aggs and successive calls were even faster as they get page cached.

sandeshkr419 commented on server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeQueryHelper.java:
```diff
@@ -146,6 +157,34 @@ private static boolean validateStarTreeMetricSupport(
+
+        if (!supportedDimensions.contains(requestField)) {
+            return false;
+        }
```
Ack. Ty, Resolved.

sandeshkr419 commented on server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/utils/StarTreeQueryHelper.java:
```diff
@@ -79,15 +80,25 @@ public static StarTreeQueryContext getStarTreeQueryContext(SearchContext context
+            }
             if (validateStarTreeMetricSupport(compositeMappedFieldType, aggregatorFactory)) {
                 continue;
             }
```
Missed earlier, rebased and fixed.

bharath-techie commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/NumericTermsAggregator.java:
```diff
@@ -145,6 +171,93 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
+
+            @Override
+            public void collectStarTreeEntry(int starTreeEntry, long owningBucketOrd) throws IOException {
+                if (valuesIterator.advance(starTreeEntry) == NO_MORE_DOCS) {
```
lets change this to !advanceExact. We've seen issues with advance when there are missing values.

bharath-techie commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/NumericTermsAggregator.java:
```diff
@@ -145,6 +171,93 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
+                    double doubleValue = ((NumberFieldMapper.NumberFieldType) context.mapperService().fieldType(fieldName)).toDoubleValue(
+                        dimensionValue
+                    );
+                    dimensionValue = NumericUtils.doubleToSortableLong(doubleValue);
```
We can revisit this later post 2.19 / probably create an issue - as we are doing 2 conversions unnecessarily.

For example if field type is float :  
1. float to double
2. double to long 
3. long to double
to fit into the existing interfaces.

[ currently as well , we're doing SortedNumericDoubleValues -> fieldData.toSortableLongBitseInts ->  NumericUtils.sortableLongToDouble for the original segment doc values , so okay to take it up post 2.19 ]

sandeshkr419 replied:
Sure - I have noted post 2.19 items in [Search Plan](https://github.com/opensearch-project/OpenSearch/issues/15257#issuecomment-2621191489) to circle back after this PR.

bharath-techie replied:
@msfroh @sandeshkr419 is there any optimizations that you can think of here ?

sandeshkr419 replied:
As a follow-up optimization, we can think of caching the fieldType - to avoid some of the data type checks. That will most likely be a minor optimization.

Alternatively, can also think of setting up a new CollectionStrategy to avoid this multiple conversions and handle the data type conversion only once during collectBucket phase.

Certainly, I think we should handle this experiment outside of this PR as I think we may have a chance to optimize the default non-star tree aggregation as well here. Thoughts @bharath-techie @msfroh?

bharath-techie replied:
Makes sense as there is scope of optimization for default and star tree aggregators.

bharath-techie reviewed (approved):
Thanks Sandesh for the changes. LGTM as changes are mostly similar to date histo aggs.

sandeshkr419 commented:
@msfroh @bharath-techie Rebased from main again since https://github.com/opensearch-project/OpenSearch/pull/17239 is merged.

msfroh commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java:
```diff
@@ -228,6 +243,10 @@ protected boolean tryPrecomputeAggregationForLeaf(LeafReaderContext ctx) throws
         }
+        CompositeIndexFieldInfo supportedStarTree = StarTreeQueryHelper.getSupportedStarTree(this.context.getQueryShardContext());
+        if (supportedStarTree != null) {
+            return preComputeWithStarTree(ctx, supportedStarTree);
```
`precomputeWithStarTree` always returns `true`.

msfroh commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java:
```diff
@@ -307,6 +326,88 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
+
+            String metricName = StarTreeUtils.fullyQualifiedFieldNameForStarTreeMetricsDocValues(
+                starTree.getField(),
+                "_doc_count",
```
Should this be hard-coded? Should it be `DocCountFieldMapper.NAME`?

msfroh commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java:
```diff
@@ -307,6 +326,88 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
+            )
+        ) {
+            @Override
+            public void setSubCollectors() throws IOException {
+                for (Aggregator aggregator : subAggregators) {
+                    this.subCollectors.add(((StarTreePreComputeCollector) aggregator).getStarTreeBucketCollector(ctx, starTree, this));
+                }
+            }
```
Can you move this below the fields defined below? Fields within an anonymous class are kind of confusing, but when they are interspersed with methods, it hurts my brain.

Actually, if you move the fields out of the anonymous class declaration and into the `getStarTreeBucketCollector` method body, they'll get captured by the anonymous `StarTreeBucketCollector` anyway (and IMO, the code would be easier to read). Maybe give that a shot and see how it looks to you.

sandeshkr419 replied:
Interesting, this was a nice suggestion actually. It works.

msfroh commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java:
```diff
@@ -307,6 +326,88 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
+        };
+    }
+
+    private boolean preComputeWithStarTree(LeafReaderContext ctx, CompositeIndexFieldInfo supportedStarTree) throws IOException {
```
Should probably return `void`, since we've already decided that we will precompute.

msfroh commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/NumericTermsAggregator.java:
```diff
@@ -145,6 +162,101 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
+    ) throws IOException {
+        assert parent == null;
+        StarTreeValues starTreeValues = StarTreeQueryHelper.getStarTreeValues(ctx, starTree);
+        return new StarTreeBucketCollector(
```
This is pretty similar to the one in `GlobalOrdinalsStringTermsAggregator`. Is there opportunity to pull the common logic into an abstract base class?

sandeshkr419 replied:
*Talking in terms of Date Histogram, Keyword Terms, Numeric Terms, Range Aggregations collectively.*

Actually it was possible to refactor it to a common utility. However, there were subtle differences in the aggregators on which it was implemented:

1. `valuesIterator` being instanceof `SortedSetStarTreeValuesIterator` or `SortedNumericStarTreeValuesIterator` in different aggregators.
2. Different valuesIterators required different handling - date being passed to rounding utilities, for range being searched in different ranges, keyword values being passed to globalOperators, numeric terms processed with correct data type conversions (as per collectionStrategy used.)
3. Bucket collection [strategy](https://github.com/opensearch-project/OpenSearch/pull/17273/files#diff-035fc79d6d7f63d9bc92823ca695de8060bde69a6a29b86aa6fa645b828a000eR525) differing in range aggregation

Based in the differences in all the aggregations, it will be possible to abstract it out by using multiple generics and consumers (or relevant biconsumers, triconusmers) - however, it will make the code highly unreadable given the code logic is similar but not same.

Instead, I'd abstract out certain small pieces, like `getDocCountIterator()`, `updateBucket()` utilities instead. Thoughts?

sandeshkr419 replied:
Abstracted out common utilities. Code looks cleaner now - check it out!

msfroh commented on server/src/main/java/org/opensearch/search/aggregations/bucket/terms/TermsAggregator.java:
```diff
@@ -291,6 +291,10 @@ private boolean subAggsNeedScore() {
     @Override
     protected boolean shouldDefer(Aggregator aggregator) {
-        return collectMode == SubAggCollectionMode.BREADTH_FIRST && !aggsUsedForSorting.contains(aggregator);
+        // don't defer when StarTreeContext is set, don't defer when collectMode == SubAggCollectionMode.BREADTH_FIRST
+        // this boolean condition can be further simplified but affects readability.
+        return (context.getQueryShardContext().getStarTreeQueryContext() == null || collectMode != SubAggCollectionMode.BREADTH_FIRST)
+            && collectMode == SubAggCollectionMode.BREADTH_FIRST
+            && !aggsUsedForSorting.contains(aggregator);
```
Could this be:

```
if (context.getQueryShardContext().getStarTreeQueryContext() == null) {
  return false;
} else {
  return collectMode == SubAggCollectionMode.BREADTH_FIRST && !aggsUsedForSorting.contains(aggregator);
}
```

That is, make it even more complicated for readability. 😁

sandeshkr419 replied:
Cool, refactored.

bharath-techie commented:
@sandeshkr419 can we revisit https://github.com/opensearch-project/OpenSearch/pull/17165#discussion_r1933025556
and https://github.com/opensearch-project/OpenSearch/pull/17165#discussion_r1933570109 . 
I think we can enable numeric terms count without any benchmarks even since there are no optimizations, we can try global aggs term count after benchmarks maybe.

sandeshkr419 commented:
Following up on https://github.com/opensearch-project/OpenSearch/pull/17165#discussion_r1933025556 

@msfroh @gbh I have added support for keyword/numeric terms aggregation without any sub-aggregations as well.

### Keyword Aggregations

To validate the performance, I indexed `clientip` as keyword field in the last of ordered dimensions to basically force a worse performance as my hunch was that the term frequency optimization in keyword terms would anytime outperform star-tree traversal, but upon benchmarking and profiling, I saw contradictory results. The terms dictionary lookup wasn't faster than star-tree traversal and the performance of star-tree (~150ms) was 2x better than term frequency optimization (~300 ms). 
The non-optimized default flow was worst as expected (~400ms). 

Based on this experimentation, I have moved up the star-tree pre-computation on higher priority than term frequency optimization. If the high cardinality field is at top in the ordered dimensions, star-tree will perform better than the current numbers, but as part of this experiment I wanted to see the worse performing scenario, which still outperforms term frequency optimization.

### Numeric Aggregations

This was a simple 10-12x improvement when using star-tree(~40ms) compared to default flow (~450ms), and since there was no other optimization in place, I simply added this case handling.


#### Benchmarking Setup

Indexed `http_logs` in Mac (M1 Pro chip, 32 GB RAM), with *G configured heap. 
Ordered dimensions in star-tree: `status`(numeric), `size` (numeric), `clientip` (keyword), `@timestamp` (date)

bharath-techie commented:
Thanks for the changes LGTM!

msfroh commented:
@sandeshkr419  -- can you please move the changelog entry from `CHANGELOG.md` to `CHANGELOG-3.0.md`?
